SAVEGAME: Reduce savegame size / complexity of save/load

Snarling opened the pull request:
Loading for Companies and Factions already initializes all companies and factions, and then just checks the savedata for valid data points and loads those data points onto the already-initialized companies and factions.

However, since both were previously using their classes + the JSONReviver for save/load, they were saving extra data (the IReviverValue), and they were also initilizing class members when loading, even though these class members would never be used as members of the class, only for loading properties onto the existing class members.

Due to no longer using the IReviverValue, the size and complexity of the savegame data is reduced, which should slightly reduce disk usage and save/load times.

Snarling committed:
Streamline saves for Company

* Company save no longer uses the reviver. The Companies object doesn't use the loaded objects directly, it just loads valid data off of the loaded objects onto statically instantiated class objects. This means the loaded objects don't actually need to be class objects.
* Also makes the savegame object simpler - it's just an object with 2 properties per company, instead of the more complicated IReviverValue that it previously was.

Snarling committed:
Same savegame changes for Factions

Same changes as previous commit, but for Factions instead of Companies.
Also some minor style changes for the new Companies stuff
Also removed some old unneeded safe loading code (not needed due to the way Factions are initialized)

Snarling committed:
Only save non-default companies

Most people don't interact with lots of companies, so often the savedata is unneeded. Check each company and only add it to save if it has non-default data.

Snarling committed:
Only save non-default data

For Factions and Companies, only save non-default data fields

Snarling committed:
Omit factions that are fully default

Going back on previous decision to always include all factions in the save. There's enough frequency of not interacting with certain factions that it may be worth omitting ones with full default values.

d0sboots commented:
Why are you changing the savegame JSON format for these?

I understand and agree with dropping everything that has default values, that saves space. And I also agree with changing *how* they're loaded, so we don't have issues with stale data if/when we change a company or faction.

But removing the IReviverValue wrapper saves minimal space, and creates the need for a migration step, which is always more risky (and has increased chance of rotting over time).

Snarling commented:
The IReviverValue overhead is not negligible when we're talking about numerous 1-layer objects with two simple properties each. The overhead is about half of each Company/Faction when it's included.

For most of game loading, we have to load objects using the constructors for the class that they're going to be loaded as, because the loaded objects directly become the game objects.

That isn't true for Faction or Company, where all actual game objects are created statically upfront, and then the savedata is just used to copy properties to the actual game objects. This allows actually typechecking the savedata, which the reviver method doesn't handle. The load method for Companies was changed in #650 and for Factions was changed in #644. There may have been some further changes since, but those were when the overall loading method was changed from "load in savedata to initialize the game objects" to "initialize game objects statically, use loaded data to copy valid properties over to game objects". This PR can really be seen as a followup on those two to make the save format actually make sense for that loading method.

d0sboots commented:
> This PR can really be seen as a followup on those two to make the save format actually make sense for that loading method.

I guess the difference in my perspective is, I don't see either of these formats as intrinsically making more or less sense for what you're doing. Rather, I see one as slightly larger but also more backward-compatible. The fact that you can simply have Reviver return the wrapped value as the migration method means it isn't *much* less backward compatible, but conversely it also shows how the wrapped format isn't any less suitable for what you're doing, either.

But since the migration is so simple, I don't have significant concerns.

Snarling committed:
Combine reviver legacy load type handling

All existing legacy load types return the same data (just unwrapped version of the IReviverValue). Instead of separately returning 5 times, have all cases converge on the same code. Details for each legacy constructor type are in comments.
